Convert Plans screen to Console page

rszwajko opened the pull request:
![Screenshot from 2023-01-13 15-36-09](https://user-images.githubusercontent.com/64194103/212353089-64a60352-38e0-4098-bd0d-1c5904e44175.png)

yaacov commented on src/modules/Plans/dynamic-plugin.ts:
```diff
@@ -55,8 +55,8 @@ export const extensions: EncodedExtension[] = [
       },
-      path: '/mtv/plans/create',
-      exact: true,
+      path: ['/mtv/plans/ns/:ns/create', '/mtv/plans/create'],
```
how do you handle a create with no namespace ?
in what namespace will the plan get created ?

rszwajko replied:
1. no namespace (aka 'All namespaces') is  handled via legacy link `/mtv/plans/create`.
2. explicit namespace is routed via `/mtv/plans/ns/:ns/create`

This is similar to what K8s paths look like  -  the namespace version has ans unique prefixe to avoid clashes (although still the longer route must go first here).

yaacov replied:
I was not clear, the question was:
when I create a new plan from "all namespaces" on what namespace will the new plan get created ?

rszwajko replied:
This will fallback to `ENV.NAMESPACE`. This is true for all other places. In case of plan wizard the logic is:
```
// TODO: add field to choose namespace if currentNamespace is 'All namepsace'
  const prefillPlanNamespace =
    (wizardMode === 'edit' ? planBeingPrefilled?.metadata?.namespace : currentNamespace) ||
    ENV.NAMESPACE;
```

yaacov replied:
lets fallback to "default" namespace for this PR, and add a bug issue to fix this when updating the wizard in next release.

rszwajko replied:
created https://github.com/kubev2v/forklift-console-plugin/issues/124

yaacov commented on src/modules/Providers/ProvidersPage.tsx:
```diff
@@ -151,10 +151,10 @@ const Page: React.FC<{
 
   return showEmptyState ? (
-    <EmptyStateProviders />
+    <EmptyStateProviders namespace={namespace} />
```
what happen when namespace is "all namespace"

rszwajko replied:
in 'All namespaces'  we have `namespace === undefined` (a legal value passed from the page props)

yaacov commented on pkg/web/src/app/queries/helpers.ts:
```diff
@@ -170,9 +170,11 @@ export const truncateK8sString = (prefix: string, suffix = '', generateName = fa
   return `${prefix.slice(0, targetPrefixLength)}${suffix}`;
 };
 
+const ALL_NAMESPECES = 'all-namespaces';
```
you can import it from the consts file ?

rszwajko replied:
This is cross package dependency (legacy code vs code in `src`).

yaacov replied:
NICE !

rszwajko commented:
Changes in the recent force push:
1. pass namespace not via `useAuthorizedK8sClient()` but only via `KubeResources` - previous version crashed in 'All namespaces' mode because `ForkliftResource` requires namespace
2. `ENV.DEFAULT_NAMESPACE === 'default'`
3. fixed a bug in WizardPanel - mapping being loaded too late
4. removed some TODOs

rszwajko commented:
Sonar related changes:
1. `useMemo` to keep stable the references passed via context - see ActionServiceDropdown
2. removed unnecessary TODOs

The remaining "smells" reported by Sonar are left unchanged:
1. arrow functions in callbacks - this impacts performance only if `React.memo()` is used for components. AFAIK the component is re-rendered if its parent is re-rendered regardless of the `useCallback()`. 
2. cognitive complexity of PlanWizard is high but refactoring at this point is out-of-scope

Suggested improvements:
1. move the "get logs" button to the kabab instead of the main-action button 
   a) follow-up issue #122 
2. add color and more letters to the providers/plans links
   a) plans use `PL` abbreviation and greenish color, providers use defaults
3. remove the types column and make it a label on the name
   a) done
4. don't show archived by default, add a selector to make them visible
   a) follow-up issue #123 
5. move source/target column to the left
   a) done
6. make description column wider
   a) added `breakWord` modifier - see below for screens with long text at various sizes
7. fix edit/duplicate actions - at some point the wizard gets blocked
   a) fixed
8. when plan fail, use the "main action" column instead of "get logs" to show a msg suggesting looking at the vms (with a link)
   a) added to #122 
9. namespace aware processing - right now the links in the legacy are static and point to "all namespaces"
   a) removed limitation to one namespace - follow-up issue #121 created to handle remaining problems

sjd78 reviewed (commented):
So. Many. Namespace. Changes.

I'll look at the rest of the files on Monday/Tuesday.  They're the actual new console page screens.  Just from skimming:
  - the data transform is daunting, just like the Providers one
  - they look pretty close in form to the Providers so my confidence is high
  - most of it will probably tracing the components down to make sure they make sense in how they're composed and how data is traveling around

rszwajko commented on pkg/web/src/app/Plans/components/VMMigrationDetails.tsx:
```diff
@@ -84,9 +85,11 @@ export type VMMigrationDetailsProps = {
 
 export const VMMigrationDetails: React.FunctionComponent<VMMigrationDetailsProps> = ({ match }) => {
-  const plansQuery = usePlansQuery();
+  const currentNamespace = match?.params?.ns;
```
The biggest challenge here was to identify classes in the legacy code base that need namespace support. Propagating the namespace via props was the easiest approach - once you change the method's signature all the invocations are not compiling which allows to easily track the impacted components.

sjd78 replied:
Ah, your wack-a-mole approach makes more sense now.

Another advantage to a hook style would be to consolidate all of the view, edit, and add namespace logic in one high-level place.  Things like which namespace to choose when creating a Provider or a Plan based on the selected namespace is in multiple places.

Could do as a refactor in the near future as well.  Actually I wouldn't want you to touch all of the namespace code within this PR again!

rszwajko commented on pkg/web/src/app/Plans/components/PlanStatusNavLink.tsx:
```diff
@@ -10,20 +10,34 @@ interface IPlanStatusNavLinkProps {
-export const PlanStatusNavLink: React.FunctionComponent<IPlanStatusNavLinkProps> = ({
-  plan,
+export const PlanNameNavLink = ({
+  name,
   isInline = true,
   children,
-}: IPlanStatusNavLinkProps) => {
+}: {
+  name: string;
+  isInline?: boolean;
+  children: React.ReactNode;
+}) => {
```
@sjd78 @yaacov 
IMHO the inferred typing offers a better dev experience - compare the following signatures presented by IDE:

![Screenshot from 2023-01-16 16-33-51](https://user-images.githubusercontent.com/64194103/212718196-93aa439b-75ad-471c-a458-d60e1b091ba6.png)

![Screenshot from 2023-01-16 16-33-22](https://user-images.githubusercontent.com/64194103/212718173-3f0fb32e-74d3-4cc5-978d-8987b5c572b1.png)

The second presents more useful info then the first one (`React.FC<>` based).

sjd78 replied:
Depends on where you mouse hover I suppose. :wink:

rszwajko commented:
# Table layout with long text at various widow sizes

## Small screen - list view displayed instead of table

![Screenshot from 2023-01-13 15-34-46](https://user-images.githubusercontent.com/64194103/212744371-01ea4335-6e12-4a20-8f0b-480490725cd0.png)

## The smallest table - just before switching to list view

![Screenshot from 2023-01-13 15-33-42](https://user-images.githubusercontent.com/64194103/212744409-7bb6a6b7-211c-4824-aeba-a8d13c81d21b.png)

## Medium sized

![Screenshot from 2023-01-13 15-33-08](https://user-images.githubusercontent.com/64194103/212744436-b490bc6d-eb6e-424c-9a73-e34125bf98cd.png)

## Standard resolution (1920x1080)
![Screenshot from 2023-01-13 15-32-12](https://user-images.githubusercontent.com/64194103/212744468-b26a1d8c-32ca-44b4-a792-954f360ff93f.png)

yaacov commented:
WOW, AMAZING :green_heart: 

nitpick: (for followup ?) I think the description column size now is too big :-) , ellipsis can help here, make the column small, but let the user see the whole text on hover, https://www.patternfly.org/v4/components/truncate#default

rszwajko commented:
@yaacov  
> ellipsis can help here, make the column small, but let the user see the whole text on hover, https://www.patternfly.org/v4/components/truncate#default
>

Actually, the text is too big:)  If we put this text into a tooltip this will be a huge tooltip.
IMHO the automatic table layout is more readable  under realistic description sizes.
We could however use the [compound expandable](https://www.patternfly.org/v4/components/table#composable-compound-expandable)  for longer descriptions.

rszwajko commented:
# Text layout with truncation

## Small screen  - normal text
![Screenshot from 2023-01-17 15-16-05](https://user-images.githubusercontent.com/64194103/212921923-d5a32137-6492-43d0-a297-a0fb941bbd45.png)
## Smallest table
![Screenshot from 2023-01-17 15-12-14](https://user-images.githubusercontent.com/64194103/212921967-f709d3c9-2d3a-424e-b4f7-bc701f63d6dc.png)
## Standard resolution
![Screenshot from 2023-01-17 15-12-59](https://user-images.githubusercontent.com/64194103/212921988-34a70e73-a312-4696-b5e4-537fab3292cd.png)
